### PR TITLE
Add support for `rust-s3` backend.

### LINKED ISSUE #11: Add support for S3 (or compatible) backend

### PR DESCRIPTION
**Pull request by lseelenbinder**
* Adds support for a S3 backend (which is optionally authenticated)

Fixes #11 and could be used to address https://github.com/maplibre/martin/discussions/1125.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -13,6 +13,8 @@ categories = ["science::geo"]
 [features]
 default = []
 http-async = ["dep:tokio", "dep:reqwest"]
+s3-async = ["dep:tokio", "dep:rust-s3", "rust-s3/tokio-native-tls"]
```
Comment (nyurik) [inline Cargo.toml:16]: maybe call it `s3-async-native` ?

File: src/error.rs [hunk 2]
```diff
@@ -32,6 +32,9 @@ pub enum PmtError {
     #[cfg(feature = "http-async")]
     #[error("{0}")]
     Http(#[from] PmtHttpError),
+    #[cfg(any(feature = "s3-async-rustls", feature = "s3-async"))]
+    #[error("{0}")]
```
Comment (nyurik) [inline src/error.rs:36]: And above as well

```suggestion
    #[error(transparent)]
```

File: src/s3.rs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                self.pmtiles_path.as_str(),
+                offset as _,
+                Some((offset + length - 1) as _),
+            )
+            .await?;
+
+        let response_bytes = response.bytes();
+
+        if response_bytes.len() > length {
+            Err(crate::error::PmtS3Error::ResponseBodyTooLong(response_bytes.len(), length).into())
```
Comment (nyurik) [inline src/s3.rs:50]: shorten it with `use ...` at the top?

**Review by nyurik — COMMENTED** [review]
looks good!  A few minor nits, and we need to re-export the underlying s3 crate somehow, but it conflicts with the `s3.rs` file.  I would recommend NOT exporting the `s3`, `http`, and `mmap` namespaces.  Instead, publish their content directly.

```
// --- this is lib.rs ---
pub use s3;  // re-export s3 crate as pmtiles::s3

mod s3_backend;
pub use s3_backend S3Backend;

mod http;
pub use http::HttpBackend;
```

File: Cargo.toml (hunk 1 above)
Comment (lseelenbinder) [reply Cargo.toml]: Good call. Done.

**Comment by lseelenbinder** [comment]
> looks good! A few minor nits, and we need to re-export the underlying s3 crate somehow, but it conflicts with the `s3.rs` file. I would recommend NOT exporting the `s3`, `http`, and `mmap` namespaces. Instead, publish their content directly.
> 
> ```
> // --- this is lib.rs ---
> 
> pub use s3;  // re-export s3 crate as pmtiles::s3
> 
> mod s3_backend;
> pub use s3_backend S3Backend;
> 
> mod http;
> pub use http::HttpBackend;
> ```

Do we want to re-export `s3` or allow users to bring their own (as we do with `reqwest`)? I tried really hard to make this TLS implementation agnostic, but `rust-s3` doesn't build properly when it doesn't have a concrete credentials type.

File: src/backend/mod.rs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#[cfg(any(feature = "s3-async-native", feature = "s3-async-rustls"))]
+pub(crate) mod s3;
+
+#[cfg(feature = "http-async")]
+pub(crate) mod http;
+
+#[cfg(feature = "mmap-async-tokio")]
+pub(crate) mod mmap;
```
Comment (nyurik) [inline src/backend/mod.rs:8]: you don't really need to control the access here - enough to say `pub mod s3` - that shouldn't make it public on the crate level, as long as you do not use `pub mod backend` in `lib.rs`.

That said, mod.rs offer a better way to compartmentalize things - you don't need to expose internals of the `backend` mod to the rest of the crate - I think it is cleaner to expose just the needed structs (up to you though) - you can also just remove the `(crate)` and keep sub-namespaces.

```suggestion
#[cfg(any(feature = "s3-async-native", feature = "s3-async-rustls"))]
mod s3;

#[cfg(any(feature = "s3-async-native", feature = "s3-async-rustls"))]
pub use s3::S3Backend;

#[cfg(feature = "http-async")]
mod http;

#[cfg(feature = "http-async")]
pub use http::HttpBackend;

#[cfg(feature = "mmap-async-tokio")]
mod mmap;

#[cfg(feature = "mmap-async-tokio")]
pub use mmap::MmapBackend;
```

**Review by nyurik — COMMENTED** [review]
> Do we want to re-export `s3` or allow users to bring their own (as we do with `reqwest`)? I tried really hard to make this TLS implementation agnostic, but `rust-s3` doesn't build properly when it doesn't have a concrete credentials type.

There is really no such thing as "bring your own" here.  If we do not re-export a dependency, users must play the magic game of "match my dependencies with the dependencies of my dependency" - i.e. user must use the same version of `reqwest` lib as used by the API of the pmtiles crate. This is an antipattern because it will only work if Cargo ends up using identical version of reqwest for both. Otherwise, it might not compile.  That's why sometimes people opt to use `reqwest = "*"` - which again is an antipattern, but they have to do it to match the API of a crate they want to use.

Instead, I think a much safer way is to either (1) ensure all pmtiles crate API only uses standard Rust types (or types declared in the pmtiles crate itself), or (2) re-export all 3rd party types together with all the other stuff related to those types - i.e. re-export `reqwest` and `s3` sub-dependencies in their own namespaces like `pmtiles::reqwest::*` and `pmtiles::s3::*`.

Also, a few inline comments.

File: src/error.rs [hunk 5]
```diff
@@ -56,3 +59,22 @@ impl From<reqwest::Error> for PmtError {
         Self::Http(PmtHttpError::Http(e))
     }
 }
+
+#[cfg(any(feature = "s3-async-rustls", feature = "s3-async-native"))]
+#[derive(Debug, Error)]
+pub enum PmtS3Error {
```
[the remark this answers is not in the document]
Comment (lseelenbinder) [reply src/error.rs]: You're right. Done.

File: src/backend/mod.rs (hunk 4 above)
Comment (lseelenbinder) [reply src/backend/mod.rs]: Nice catch, cleaned up.

**Comment by lseelenbinder** [comment]
> There is really no such thing as "bring your own" here. If we do not re-export a dependency, users must play the magic game of "match my dependencies with the dependencies of my dependency" - i.e. user must use the same version of reqwest lib as used by the API of the pmtiles crate. This is an antipattern because it will only work if Cargo ends up using identical version of reqwest for both. Otherwise, it might not compile. That's why sometimes people opt to use reqwest = "*" - which again is an antipattern, but they have to do it to match the API of a crate they want to use.
> 
> Instead, I think a much safer way is to either (1) ensure all pmtiles crate API only uses standard Rust types (or types declared in the pmtiles crate itself), or (2) re-export all 3rd party types together with all the other stuff related to those types - i.e. re-export reqwest and s3 sub-dependencies in their own namespaces like pmtiles::reqwest::* and pmtiles::s3::*.

Fair enough. And users can always bring their own, if they want—just because we're nice and re-export doesn't mean users will use the exports. 😄

**Review by nyurik — APPROVED** [review]
thanks!!

**Pull request merged**